packages/main/src/plugin/container-registry.ts

```diff
@@ -1347,6 +1391,135 @@ export class ContainerProviderRegistry {
     }
   }
 
+  async attachContainer(
+    engineId: string,
+    containerId: string,
+    onData: (data: string) => void,
+    onError: (error: string) => void,
+    onEnd: () => void,
+  ): Promise<(param: string) => void> {
+    // check if we have an existing stream
+    let attachStream = this.streamsPerContainerId.get(containerId);
+
+    const setupStream = (stream: NodeJS.ReadWriteStream) => {
+      stream.on('data', chunk => {
+        onData(chunk.toString('utf-8'));
+      });
+
+      stream.on('error', err => {
+        onError(String(err));
+      });
+
+      stream.on('end', () => {
+        onEnd();
+      });
+
+      // do we have previous data?
+      const previousData = this.streamsOutputPerContainerId.get(containerId);
+      if (previousData) {
+        const concat = Buffer.concat(previousData);
+        // replay
+        onData(concat.toString('utf-8'));
+      }
+    };
+
+    if (!attachStream) {
+      // grab the container object
+      const container = this.getMatchingContainer(engineId, containerId);
+
+      const getAttachStream = async () => {
+        // use either podman specific API or compat API
+        try {
+          const libpod = this.getMatchingPodmanEngine(engineId);
+          return libpod.podmanAttach(container.id);
+        } catch (error) {
+          // run attach
+          const compatAttachOptions: ContainerAttachOptions = {
+            stream: true,
+            stdin: true,
+            stdout: true,
+            stderr: true,
+            hijack: true,
+          };
+          return container.attach(compatAttachOptions);
+        }
+      };
+
+      getAttachStream()
+        .then(readWriteStream => {
+          attachStream = readWriteStream;
+          setupStream(attachStream);
+        })
+        .catch((err: unknown) => {
+          console.log('error in attach', err);
+        });
+    } else {
+      setupStream(attachStream);
+    }
+
+    this.telemetryService.track('attachContainer');
+
+    return (param: string) => {
+      attachStream?.write(param);
+    };
+  }
+
+  // keep a reference to the input/output stream of a container
+  // can be used before starting the container
+  // it only keep data if tty is specified on the container
+  async attachToContainer(
+    engine: InternalContainerProvider,
+    container: Dockerode.Container,
+    hasTty?: boolean,
+    openStdin?: boolean,
+  ): Promise<void> {
+    // if option is not specified, try to look if the container is using tty or not
+    if (hasTty === undefined || openStdin === undefined) {
+      const containerInspectInfo = await container.inspect();
+      hasTty = containerInspectInfo.Config.Tty;
+      openStdin = containerInspectInfo.Config.OpenStdin;
+    }
+    // no tty and no stdin, do not need to try to attach a terminal
+    if (!hasTty && !openStdin) {
```

Comment:
Should the case when hasTty = true and openStdin=false be also banned? It is kind of useless to have terminal enabled without any input attached. Nothing is going to show up in there on itself and no reaction on input ether.

Is there a PR for renderer part to test?

Reply:
I've pushed a fixup commit as yes we should try to keep data only if it is useful.

so, here in this code, we were just trying to attach the output after the start of a container but anyway we were not displaying a way to connect. But better to not keep data if we're not using them.

Reply:
here is the PR to test the renderer part https://github.com/containers/podman-desktop/pull/3900